tools/tpm2_tool.h

```diff
@@ -33,36 +33,54 @@
 
 #include <sapi/tpm20.h>
 #include <stdbool.h>
-#include "options.h"
+
+#include "tpm2_options.h"
 
 extern bool output_enabled;
 
-int
-execute_tool (int              argc,
-              char             *argv[],
-              char             *envp[],
-              common_opts_t    *opts,
-              TSS2_SYS_CONTEXT *sapi_context);
+/**
+ * An optional interface for tools to specify what options they support.
+ * They are concatenated with main's options and passed to getopt_long.
+ * @param opts
+ *  The callee can choose to set *opts to a tpm_options pointer allocated
+ *  via tpm2_options_new(). Setting *opts to NULL is not an error, and
+ *  Indicates that no options are specified by the tool.
+ *
+ * @return
+ *  True on success, false on error.
+ */
+bool tpm2_tool_onstart(tpm2_options **opts) __attribute__((weak));
 
-/*
- * Prints output messages if not disabled. Output messages are what tools prints
- * to the standard output during normal execution.
+/**
+ * This is the main interface for tools, after tcti and sapi initialization
+ * are performed.
+ * @param sapi_context
+ *  The system api context.
+ * @param flags
+ *  Flags that tools may wish to respect.
+ * @return
+ *  0 on success.
+ */
+int tpm2_tool_onrun (TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) __attribute__((weak));
+
+/**
+ * Called when the tool is exiting, useful for cleanup.
+ */
+void tpm2_tool_onexit(void) __attribute__((weak));
+
+/**
+ * prints output to stdout respecting the quiet option.
+ * Ie when quiet, don't print.
+ * @param fmt
+ *  The format specifier, ala printf.
+ * @param ...
+ *  The varargs, just like printf.
  */
-#define TOOL_OUTPUT(fmt, ...)                   \
+#define tpm2_tool_output(fmt, ...)                   \
```

Comment:
I would do this as a separate patch.

Reply:
Normally I would be all about keeping patches as small as possible, with everything separated out. However, with this gigantic change, I am less worried about granularity.

If I just write our own option parser for main(), we can get away leaving everything else, I am very tempted to go in that direction. That would be the smallest, simplest change to fix our option issue.

Reply:
Do you mean not using getopt() ? I'm not sure about that, writing correct/safe C code is already hard and parsers are specially prone to subtle errors. I'm obviously not talking about you, just saying in general.